src/accumulator/proof.rs

```diff
@@ -0,0 +1,310 @@
+    ///   // 08       09       10       11         
+    ///   // |----\   |----\   |----\   |----\       
+    ///   // 00   01  02   03  04   05  06   07
+    ///   // For proving 0, we need 01, 09 and 13's hashes. 00, 08, 12 and 14 can be calculated
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7").unwrap());
+    ///   
+    ///   let mut hashes = Vec::new();
+    ///   for i in test_values {
+    ///     let mut engine = bitcoin_hashes::sha256::Hash::engine();
+    ///      engine.input(&[i]);
+    ///      let hash = sha256::Hash::from_engine(engine);
+    ///     hashes.push(hash);
+    ///   }
+    ///
+    ///   s.modify(&hashes, &vec![]);
+    ///   let p = Proof::new(targets, proof_hashes);
+    ///   assert!(p.verify(&vec![hashes[0]] , &s).expect("No error should happens"));
+    ///```
+    pub fn verify(&self, del_hashes: &Vec<sha256::Hash>, stump: &Stump) -> Result<bool, String> {
+        let calculated_roots = self.create_root_candidates(del_hashes, stump)?;
+        let stump_roots = util::num_roots(stump.leafs);
+        let mut number_matched_roots = 0;
+
+        for i in 0..stump_roots {
+            if calculated_roots[number_matched_roots] == stump.roots[stump.roots.len() - (i + 1)] {
+                number_matched_roots += 1;
+                if number_matched_roots == calculated_roots.len() {
+                    break;
+                }
+            }
+        }
+
+        if calculated_roots.len() != number_matched_roots && calculated_roots.len() != 0 {
+            return Ok(false);
+        }
+        Ok(true)
+    }
+
+    pub fn create_root_candidates(
+        &self,
+        del_hashes: &Vec<sha256::Hash>,
+        stump: &Stump,
+    ) -> Result<Vec<sha256::Hash>, String> {
+        // Where all the root hashes that we've calculated will go to.
+        let total_rows = util::tree_rows(stump.leafs);
+        let mut targets = self.targets.clone();
+
+        // Where all the parent hashes we've calculated in a given row will go to.
+        let mut calculated_root_hashes =
+            Vec::<sha256::Hash>::with_capacity(util::num_roots(stump.leafs) as usize);
+
+        // As we calculate nodes upwards, it accumulates here
+        let mut nodes: Vec<_> = self
+            .targets
+            .to_owned()
+            .into_iter()
+            .zip(del_hashes.to_owned())
+            .collect();
+
+        let mut row = 0;
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
```

Comment:
`detect_row_hashes()` will only work if `targets` are sorted. Since there's no requirement that the caller has to pass in the targets as sorted, we should be sorting the targets before we use them.

After we declare `nodes` we should be sorting both the nodes and the targets before using them.

```
      "leafs":[
         0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
        20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 
        40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59,
        60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 
        80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99
       ],
       "proof": [
         "e7cf46a078fed4fafd0b5e3aff144802b853f8ae459a4f0c14add3314b7cc3a6",
         "f299791cddd3d6664f6670842812ef6053eb6501bd6282a476bbbf3ee91e750c",
         "2f0fd1e89b8de1d57292742ec380ea47066e307ad645f5bc3adad8a06ff58608",
         "bb7208bc9b5d7c04f1236a82a0093a5e33f40423d5ba8d4266f7092c3ba43b62",
         "3e23e8160039594a33894f6564e1b1348bbd7a0088d42c4acb73eeaed59c009d",
         "cd9c77062a338e63a63ca623db438cb8676f15466641079ee61ec2dda98de796",
         "83b2d85c9ae3143b1b1598076c85b3830d7f35a6c67c6ad122df8ab6235c40a1",
         "60e6d84b9ad3355211af775aff3338289431744c79fc06f5e2c065cbdd690f79",
         "9a899064135333436906fe5b714ca017e61d9e952860dcda410dec4bd0fd59a5",
         "18b41f49a19d41613b4f717b3ef291e2cde3b963fcf60f9a371affd761f1b998",
         "46c5f8e5b7ccdb3eb2014f988c30a73ba8bacfccf282293acbe559e56d540a17",
         "e799acb98a071c4884707e4bc8c093ba22571c8d84cc0223ab0c2c9327313a5b",
         "6faf95884aaf779221fbe2f81d68bbac26a70751d384fa5831b2d6d0cf07293d",
         "a8fdcb2277476c5a76bf6a5e1d107f5c48ec59a7c813f6f9f1419a648e5568af",
         "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42",
         "94b8dc9cecafe156b80e4e64414cf4204858f56713776c525c38fb14c2c5e622",
         "8ec4ba1ba4d169b8a1e5261408d9799185d984206b1d8779e4a5c1e431023fc1"
       ],
       "targets": [
         10, 19, 20, 30, 31, 32, 99 
       ],
       "expected": true
     },
```

When we change the targets to:

```
20, 19, 10, 30, 31, 32, 99
```

the test will fail.

Reply:
I was working with the assumption that `targets` is sorted. Sorting `targets` also requires permutations on `hashes`, right? We assume that a node's hash is at the same index as the node itself.

Reply:
Yes we expect the caller to give the `hashes` in the same permutation as the `targets`. When we sort the targets, we should be changing the ordering of the `hashes` as well.

Reply:
Just extrapolated the iterator idea, I think the version in [4ffc6c1](https://github.com/mit-dci/rustreexo/pull/2/commits/4ffc6c1ff9f5324dad985c6e637f13500041b5f3) is way more readable and safe than before. It also sorts at the beginning, so I think we don't have this bug anymore.  

Unrelated: CI is working :tada: . I made a basic integration with rustfmt in my repo, will test it here.